### PR TITLE
Feat/pps

### PR DESCRIPTION
**Pull request by TwistingTwists**

### CODE REVIEW

File: lib/fluid/model/resources/tag/changes/uct2suctorucp.ex [hunk 1]
```diff
@@ -12,13 +12,34 @@ defmodule Fluid.Model.Warehouse.Changes.UCT2SUCTorUCP do
...
+      when pool_capacity in [:fixed, :capped] ->
+        # if tank or pool are not uncapped, halt the pipeline and let it pass.
+        {:halt, cs}
+
+      _ ->
+        {:cont, cs}
+    end
+  end
+
+  def src_dest_from_different_wh({:cont, changeset}, source, dest) do
```
Comment (thevole) [inline lib/fluid/model/resources/tag/changes/uct2suctorucp.ex:42]: Can this function be private?

File: lib/fluid/model/resources/tag/changes/uct2suctorucp.ex [hunk 2]
```diff
@@ -31,15 +52,15 @@ defmodule Fluid.Model.Warehouse.Changes.UCT2SUCTorUCP do
...
-      {:ok, changeset}
+      {:cont, changeset}
     end
   end
 
-  # def src_dest_from_different_wh({:error, changeset}, _source, _dest), do: {:error, changeset}
+  def src_dest_from_different_wh({:halt, changeset}, _source, _dest), do: {:halt, changeset}
 
-  def src_dest_type_validations({:ok, changeset}, source, dest) do
+  def src_dest_type_validations({:cont, changeset}, source, dest) do
```
Comment (thevole) [inline lib/fluid/model/resources/tag/changes/uct2suctorucp.ex:63]: Private?

File: lib/fluid/model/resources/warehouse/calculations/pool_tank_type.ex [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  @impl true
+  def init(opts) do
+    {:ok, opts}
+  end
+
+  @impl true
+  def select(_, opts, _) do
+    [opts[:field]]
+  end
```
Comment (thevole) [inline lib/fluid/model/resources/warehouse/calculations/pool_tank_type.ex:16]: What is the module's responsibility? I'm not sure of your approach here.

File: lib/fluid/model/resources/pool.ex [hunk 4]
```diff
@@ -2,7 +2,8 @@ defmodule Fluid.Model.Pool do
   require Logger
 
   use Ash.Resource,
-    data_layer: AshPostgres.DataLayer
+    data_layer: AshPostgres.DataLayer,
+    extensions: [AshJason.Extension]
```
Comment (thevole) [inline lib/fluid/model/resources/pool.ex:6]: Why use Jason now? What's the thinking here?

File: lib/fluid/model/resources/pps.ex [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    {:ok,
+     Enum.map(pps_list, fn pps ->
+       pools = Kernel.get_in(pps, Enum.map(fields, &Access.key/1))
+
+       Enum.map(pools, fn %{warehouse_id: wh_id} ->
+         Map.get(all_wh_hashmap, wh_id, nil)
+       end)
+       |> Enum.uniq()
+     end)}
```
Comment (thevole) [inline lib/fluid/model/resources/pps.ex:104]: Can the inner fn be refactored into a defp with a clear name?

File: lib/fluid/model/resources/tank.ex [hunk 6]
```diff
@@ -2,7 +2,8 @@ defmodule Fluid.Model.Tank do
   require Logger
 
   use Ash.Resource,
-    data_layer: AshPostgres.DataLayer
+    data_layer: AshPostgres.DataLayer,
+    extensions: [AshJason.Extension]
```
Comment (thevole) [inline lib/fluid/model/resources/tank.ex:6]: Why use Jason here?

File: lib/fluid/model/resources/pool.ex (hunk 4 above)
Comment (TwistingTwists) [reply lib/fluid/model/resources/pool.ex:6]: we are serialising the entire struct as map to store in db. currently, this is used in `Tag` module to store `source` and `destination`. 

storing data in this de-normalised way avoid excessive db calls. This way, `Tag` has `source` and `destination` whose data is stored as map and can be accesssed directly.

File: lib/fluid/model.ex [hunk 7]
```diff
@@ -112,4 +123,104 @@ defmodule Fluid.Model do
...
+    Enum.reduce(all_tags, {cts_acc, tags_acc}, fn tag, {cts_acc, tags_acc} ->
+      if tag.source["id"] == pool.id do
+        # assumption that tag.destination is ct
+        ct = Model.Tank.read_by_id!(tag.destination["id"])
+        {[ct | cts_acc], [tag | tags_acc]}
+      else
+        {cts_acc, tags_acc}
+      end
+    end)
+  end
```
Comment (thevole) [inline lib/fluid/model.ex:225]: Maybe refactor inner fn into named defp?

File: test/fluid/pps/unit_tests/overlap_algo_test.exs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    assert pool_overlap_map == %{1 => [0], 2 => [0], 3 => [0]}
+    assert merged_lists == [[%{name: "pool1", id: 1}, %{name: "pool2", id: 2}, %{name: "pool3", id: 3}]]
+  end
+
+  test "pool_overlap_map with multiple pool lists - forming multiple PPS - same as original list" do
+    # all_pools_list represents the pools that maybe form a pps. these are calculations per capped tank.
+    # so, the overlapping pools need to be merged to create a final pps.
+    all_pools_list = [
+      [%{name: "pool1", id: 1}, %{name: "pool2", id: 2}],
```
Comment (thevole) [inline test/fluid/pps/unit_tests/overlap_algo_test.exs:35]: Not sure of the terminology here. What does 'overlap' mean?

**Review by thevole — CHANGES_REQUESTED** [review]
Left some comments and questions.

File: lib/fluid/model/resources/tag/changes/uct2suctorucp.ex (hunk 1 above)
Comment (TwistingTwists) [reply lib/fluid/model/resources/tag/changes/uct2suctorucp.ex:42]: yes

File: test/fluid/pps/unit_tests/overlap_algo_test.exs (hunk 8 above)
Comment (TwistingTwists) [reply test/fluid/pps/unit_tests/overlap_algo_test.exs:41]: added docs

File: lib/fluid/model/resources/pps.ex (hunk 5 above)
Comment (TwistingTwists) [reply lib/fluid/model/resources/pps.ex]: done

**Pull request merged**